### PR TITLE
WebBrowserSpec refactor

### PR DESCRIPTION
**Pull request by aweng98**
**PR checklist**
- [ ] For each library you've modified here, decide whether it requires a major, minor, or no version bump. (Click [here](/broadinstitute/workbench-libs/blob/develop/CONTRIBUTING.md) for guidance)

If you're doing a **major** or **minor** version bump to any libraries:
- [ ] Bump the version in `project/Settings.scala` `createVersion()`
- [ ] Update `CHANGELOG.md` for those libraries
- [ ] I promise I used `@deprecated` instead of deleting code where possible

In all cases:
- [ ] Replace the appropriate version hashes in `README.md` and the `CHANGELOG.md` for any libs you changed with `TRAVIS-REPLACE-ME` to auto-update the version string
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green (CI _and_ coverage tests)
- [ ] Squash commits and **merge to develop**
- [ ] Delete branch after merge

### CODE REVIEW

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 1]
```diff
@@ -22,7 +22,15 @@ import scala.sys.SystemProperties
   */
 trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogging with RandomUtil { self: Suite =>
 
-  val api = Orchestration
+  lazy val api = Orchestration
+  lazy val headless = new SystemProperties().get("headless")
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala]: evaluated once

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 2]
```diff
@@ -22,7 +22,15 @@ import scala.sys.SystemProperties
...
+
+  val isHeadless: Boolean = {
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala:28]: val is evaluated once

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 3]
```diff
@@ -43,25 +51,26 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
...
   }
 
-  private def getChromeIncognitoOption(downloadPath: String, headless: Boolean): ChromeOptions = {
+  private def getChromeIncognitoOption(downloadPath: String): ChromeOptions = {
     val fullDownloadPath = new File(downloadPath).getAbsolutePath
     logger.info(s"Chrome download path: $fullDownloadPath")
     val options = new ChromeOptions
     options.addArguments("--incognito")
     options.addArguments("--no-experiments")
+    // https://github.com/GoogleChrome/chrome-launcher/blob/master/docs/chrome-flags-for-tools.md
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala]: enhance Chrome for running tests

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 4]
```diff
@@ -72,12 +81,13 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
     options
   }
 
-  private def runLocalChrome(options: ChromeOptions, testCode: (WebDriver) => Any): Unit = {
-    val service = new ChromeDriverService.Builder().usingDriverExecutable(new File(Config.ChromeSettings.chromeDriverPath)).usingAnyFreePort().build()
+  lazy val chromeDriverHost: String = Config.ChromeSettings.chromedriverHost
+  lazy val chromeDriverFile: File = new File(Config.ChromeSettings.chromeDriverPath)
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala:86]: evaluated once

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 5]
```diff
@@ -111,7 +126,7 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
       f
     } catch {
       case t: Throwable =>
-        val date = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss-SSS").format(new java.util.Date())
+        val date = new SimpleDateFormat("HH-mm-ss-SSS").format(new java.util.Date())
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala]: shorter filename

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 6]
```diff
@@ -133,7 +148,9 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
             val logString = logLines.map(_.toString).reduce(_ + "\n" + _)
             new FileOutputStream(new File(logFileName)).write(logString.getBytes)
           }
+          logger.error(s"Screenshot ${date}_$suiteName.png Exception. ", t)
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala]: log exception to test log

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 7]
```diff
@@ -111,10 +129,11 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
       f
     } catch {
       case t: Throwable =>
-        val date = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss-SSS").format(new java.util.Date())
-        val fileName = s"failure_screenshots/${date}_$suiteName.png"
-        val htmlSourceFileName = s"failure_screenshots/${date}_$suiteName.html"
-        val logFileName = s"failure_screenshots/${date}_${suiteName}_console.txt"
+        val date = new SimpleDateFormat("HH-mm-ss-SSS").format(new java.util.Date())
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala:130]: existing filename format is alittle long. I'd like to shorten it. New filename format has `suiteName` in front.
`2018-06-04_19-33-13-787_DataSpec.png` vs `DataSpec_21-45-05-248.png`

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 8]
```diff
@@ -31,7 +39,7 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
     *
     * @param testCode the test code to run
     */
-  def withWebDriver(testCode: (WebDriver) => Any): Unit = {
+  def withWebDriver(testCode: WebDriver => Any): Unit = {
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala:42]: unnecessary parentheses

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserUtil.scala [hunk 9]
```diff
@@ -9,7 +9,8 @@ import scala.collection.JavaConverters._
   * Mix-in utilities for ScalaTest's WebBrowser.
   */
 trait WebBrowserUtil extends WebBrowser {
-  val defaultTimeOutInSeconds: Long = 30
+
+  val defaultTimeOutInSeconds: Long = 40
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserUtil.scala:13]: Chrome UI is slower in parallel test execution

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 10]
```diff
@@ -43,25 +51,28 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
...
     val options = new ChromeOptions
     options.addArguments("--incognito")
     options.addArguments("--no-experiments")
+    // https://github.com/GoogleChrome/chrome-launcher/blob/master/docs/chrome-flags-for-tools.md
+    options.addArguments("--disable-background-networking")
+    options.addArguments("--disable-client-side-phishing-detection")
+    options.addArguments("--test-type")
+    options.addArguments("--enable-automation")
+    options.addArguments("--no-sandbox")
+    options.addArguments("--disable-extensions")
```
Comment (aweng98) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala]: chrome switches to try out

File: serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala [hunk 11]
```diff
@@ -43,25 +51,28 @@ trait WebBrowserSpec extends WebBrowserUtil with ExceptionHandling with LazyLogg
...
     val fullDownloadPath = new File(downloadPath).getAbsolutePath
     logger.info(s"Chrome download path: $fullDownloadPath")
     val options = new ChromeOptions
     options.addArguments("--incognito")
     options.addArguments("--no-experiments")
+    // https://github.com/GoogleChrome/chrome-launcher/blob/master/docs/chrome-flags-for-tools.md
+    options.addArguments("--disable-background-networking")
+    options.addArguments("--disable-client-side-phishing-detection")
+    options.addArguments("--test-type")
+    options.addArguments("--enable-automation")
```
Comment (breilly2) [inline serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala]: I suspect this argument has been on by default in our environment. I've seen the "Chrome is being controlled by automated test software" message when running tests.

Comment (aweng98) [reply serviceTest/src/test/scala/org/broadinstitute/dsde/workbench/service/test/WebBrowserSpec.scala]: I'll remove it.

**Review by gdlex4015 — APPROVED** [review]
If all checks pass ok

**Pull request merged**